### PR TITLE
Refactoring structure

### PR DESCRIPTION
**Pull request by ctlong**
- Made AddressGeocoder class into module and moved functions in to AddressGeocoder::Base class
- Split many functions off into Parse, Url, and Request classes
- Added versions to dependencies
- Refactored using rubocop.

### CODE REVIEW

File: lib/address_geocoder/url.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+require 'uri'
+
+class Url # :nodoc:
+  GOOGLE_TITLES = {
+    country: 'country',
+    postal_code: 'postal_code',
+    city: 'locality',
+    state: 'administrative_area'
+  }.freeze
```
Comment (wingleungchoi) [inline lib/address_geocoder/url.rb]: why freeze it?

File: lib/address_geocoder/parse.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    @suggested_address.delete(:switch)
+    @suggested_address
+  end
+
+  def parse_field(field)
+    @field = field
+    if contains?(CITY_TYPES)
+      matched_city(field['long_name'])
+    elsif contains?(CITY_TYPES)
+      matched_state(field['long_name'])
```
Comment (wingleungchoi) [inline lib/address_geocoder/parse.rb]: line 21, 22 are duplicated with line 19, 20

File: lib/address_geocoder/parse.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    if @suggested_address[:city]
+      @suggested_address[:state]  = value
+      @suggested_address[:switch] = true
+    else
+      @suggested_address[:city] = value
+    end
+  end
+
+  def matched_state(value)
+    if switched?
```
Comment (wingleungchoi) [inline lib/address_geocoder/parse.rb]: interesting :smile:

**Comment by wingleungchoi** [comment]
Just leave some comments. The new structure is more beautiful.
Should we use `module AddressGeocoder` or `class AddressGeocoder ` for each module prefix?

For example, the gem savon using `class Sekken` 
https://github.com/savonrb/sekken/blob/058fd2784e139a8d0261b666c697da284eaa2a04/lib/sekken/errors.rb

File: lib/address_geocoder/url.rb (hunk 1 above)
Comment (ctlong) [inline lib/address_geocoder/url.rb]: Correct, we don't need to change that variable.

**Comment by ctlong** [comment]
@wingleungchoi I like the sekken way better, that way we wouldn't have to call AddressGeocoder::Base right? Just AddressGeocoder.

**Comment by wingleungchoi** [comment]
sekken way looks cool.
It uses ruby one of  characteristics the dynamic script. #=> in the following example
https://github.com/savonrb/sekken/blob/058fd2784e139a8d0261b666c697da284eaa2a04/lib/sekken.rb
each `require` like `require 'sekken/errors'`, we might it is something like the following
```
#require 'sekken/errors' #=> syntax, the opening 
class Sekken #(open/create a class named sekken)
  # define error
end

require 'sekken/operation'
class Sekken
   class Operation  #(under Sekken (Constant/Namespace) create the another class Operation)
        # define methods
        # remind Operation class have NOOOOOOO inheritance  relationship
       http://stackoverflow.com/questions/14739640/ruby-classes-within-classes-or-modules-within-modules#answer-14740440
   end
end

Finally in line 10 of https://github.com/savonrb/sekken/blob/058fd2784e139a8d0261b666c697da284eaa2a04/lib/sekken.rb
class Sekken #(Open class again and add more methods inside)
    #add more methods inside
end
```

**Comment by wingleungchoi** [comment]
@ctlong 
reopen class in ruby
FYI http://rubylearning.com/satishtalim/ruby_open_classes.html
reopen class => monkey patch http://culttt.com/2015/06/17/what-is-monkey-patching-in-ruby/
Good way to monkey - patching http://www.justinweiss.com/articles/3-ways-to-monkey-patch-without-making-a-mess/

**Pull request merged**